File: Algorithm/Rolling Hash/atcoder/B56/GPT/B56.go
// では、先ほどの **二重ローリングハッシュ法による回文判定** を Go 1.20.6 で実装します。

// * 型はすべて明示
// * 関数ごとにパラメータ・返却値をコメントで記載
// * `time` パッケージで処理時間計測
// * `runtime` パッケージでメモリ使用量計測
// * 入出力は高速化のために `bufio` を使用

// ---

// ```go
// Go 1.20.6
// 二重ローリングハッシュによる高速回文判定
// 実行方法: go run main.go < input.txt

package main

import (
	"bufio"
	"fmt"
	"os"
	"runtime"
	"time"
)

// buildHash は文字列 S に対して基数 base と法 mod を用いた
// 累乗配列、正方向ハッシュ配列、逆方向ハッシュ配列を構築する
//
// Parameters:
//   S    string : 対象文字列（英小文字）
//   base int64  : ローリングハッシュの基数
//   mod  int64  : 法（素数）
//
// Returns:
//   pow []int64 : base^i % mod の累乗配列
//   hf  []int64 : 正方向 prefix hash 配列（1-indexed）
//   hr  []int64 : 逆方向 prefix hash 配列（1-indexed）
func buildHash(S string, base, mod int64) ([]int64, []int64, []int64) {
	n := len(S)
	pow := make([]int64, n+1)
	hf := make([]int64, n+1)
	hr := make([]int64, n+1)

	pow[0] = 1
	for i := 1; i <= n; i++ {
		pow[i] = (pow[i-1] * base) % mod
	}

	for i := 0; i < n; i++ {
		val := int64(S[i]-'a'+1)
		hf[i+1] = (hf[i]*base + val) % mod
	}

	for i := 0; i < n; i++ {
		val := int64(S[n-1-i]-'a'+1)
		hr[i+1] = (hr[i]*base + val) % mod
	}

	return pow, hf, hr
}

// getSubHash は prefix hash 配列から部分文字列 [l, r] (1-indexed) のハッシュ値を取得する
//
// Parameters:
//   hf   []int64 : prefix hash 配列（1-indexed）
//   pow  []int64 : base^i % mod の累乗配列
//   l    int     : 部分文字列の開始位置 (1-indexed)
//   r    int     : 部分文字列の終了位置 (1-indexed)
//   mod  int64   : 法
//
// Returns:
//   hashValue int64 : 部分文字列のハッシュ値
func getSubHash(hf, pow []int64, l, r int, mod int64) int64 {
	length := r - l + 1
	res := hf[r] - (hf[l-1]*pow[length])%mod
	res %= mod
	if res < 0 {
		res += mod
	}
	return res
}

// solve は N, Q, S, クエリ配列を受け取り回文判定結果を返す
//
// Parameters:
//   N       int           : 文字列長
//   Q       int           : クエリ数
//   S       string        : 対象文字列
//   queries [][2]int      : 各クエリ (L, R) の配列
//
// Returns:
//   results []string : 各クエリに対する "Yes" または "No" の結果
func solve(N, Q int, S string, queries [][2]int) []string {
	const MOD1 int64 = 1000000007
	const MOD2 int64 = 1000000009
	const BASE1 int64 = 1000003
	const BASE2 int64 = 1000033

	pow1, hf1, hr1 := buildHash(S, BASE1, MOD1)
	pow2, hf2, hr2 := buildHash(S, BASE2, MOD2)

	results := make([]string, Q)
	for i, q := range queries {
		L, R := q[0], q[1]

		fh1 := getSubHash(hf1, pow1, L, R, MOD1)
		fh2 := getSubHash(hf2, pow2, L, R, MOD2)

		revL := N - R + 1
		revR := N - L + 1

		rh1 := getSubHash(hr1, pow1, revL, revR, MOD1)
		rh2 := getSubHash(hr2, pow2, revL, revR, MOD2)

		if fh1 == rh1 && fh2 == rh2 {
			results[i] = "Yes"
		} else {
			results[i] = "No"
		}
	}
	return results
}

func main() {
	in := bufio.NewReaderSize(os.Stdin, 1<<20)
	out := bufio.NewWriterSize(os.Stdout, 1<<20)
	defer out.Flush()

	var N, Q int
	fmt.Fscan(in, &N, &Q)

	var S string
	fmt.Fscan(in, &S)

	queries := make([][2]int, Q)
	for i := 0; i < Q; i++ {
		fmt.Fscan(in, &queries[i][0], &queries[i][1])
	}

	start := time.Now()
	results := solve(N, Q, S, queries)
	elapsed := time.Since(start)

	for _, res := range results {
		fmt.Fprintln(out, res)
	}

	// メモリ使用量計測
	var m runtime.MemStats
	runtime.ReadMemStats(&m)

	fmt.Fprintf(os.Stderr, "Time: %v ms\n", elapsed.Milliseconds())
	fmt.Fprintf(os.Stderr, "Memory: %.2f MB\n", float64(m.Alloc)/1024.0/1024.0)
}
// ```

// ---

// ### 実装の特徴

// 1. **二重ローリングハッシュ**

//    * `MOD1, MOD2` と `BASE1, BASE2` を変えて衝突率をほぼゼロに。
//    * Go の `int64` 型を使用して安全に計算。

// 2. **関数分割**

//    * `buildHash` → 累乗配列・正逆ハッシュ配列生成
//    * `getSubHash` → 部分文字列の O(1) ハッシュ取得
//    * `solve` → クエリを処理して結果を返す

// 3. **高速入出力**

//    * `bufio.NewReaderSize` / `bufio.NewWriterSize` で大きめバッファを確保。

// 4. **計測**

//    * `time.Since()` で経過時間計測
//    * `runtime.ReadMemStats` でメモリ使用量取得

File: Algorithm/Rolling Hash/atcoder/B56/GPT/B56.py
# 先ほどの **二重ローリングハッシュ + BigInt（PythonではintでOK）** の実装を、Python 3.11.4 向けに型付きで書き直します。
# `sys.stdin.buffer` で高速入力、`time` と `resource` で処理時間・メモリ計測を行います。
# すべて関数化し、パラメータと戻り値の型をコメントで記載します。

# ---

# ```python
# Python 3.11.4
# 高速回文判定 (二重ローリングハッシュ法)
# 実行方法: python main.py < input.txt

from __future__ import annotations
import sys
import time
import resource
from typing import List, Tuple

# =========================
# 関数群
# =========================

def build_hash(S: str, base: int, mod: int) -> Tuple[List[int], List[int], List[int]]:
    """
    与えられた文字列 S に対して、基数 base と法 mod を用いた
    正方向・逆方向のハッシュ配列と累乗配列を構築する。

    Parameters
    ----------
    S : str
        対象文字列（英小文字）
    base : int
        ローリングハッシュの基数
    mod : int
        法（素数）

    Returns
    -------
    pow_list : List[int]
        base^i % mod の配列（0-indexed, 長さ N+1）
    hf : List[int]
        正方向 prefix hash 配列（1-indexed）
    hr : List[int]
        逆方向 prefix hash 配列（1-indexed）
    """
    n: int = len(S)
    pow_list: List[int] = [0] * (n + 1)
    hf: List[int] = [0] * (n + 1)
    hr: List[int] = [0] * (n + 1)

    pow_list[0] = 1
    for i in range(1, n + 1):
        pow_list[i] = (pow_list[i - 1] * base) % mod

    for i in range(n):
        val = ord(S[i]) - 96
        hf[i + 1] = (hf[i] * base + val) % mod

    for i in range(n):
        val = ord(S[n - 1 - i]) - 96
        hr[i + 1] = (hr[i] * base + val) % mod

    return pow_list, hf, hr


def get_sub_hash(hf: List[int], pow_list: List[int], l: int, r: int, mod: int) -> int:
    """
    prefix hash 配列から部分文字列 [l, r] (1-indexed) のハッシュ値を取得する。

    Parameters
    ----------
    hf : List[int]
        prefix hash 配列（1-indexed）
    pow_list : List[int]
        base^i % mod の累乗配列
    l : int
        部分文字列の開始位置 (1-indexed)
    r : int
        部分文字列の終了位置 (1-indexed)
    mod : int
        法

    Returns
    -------
    hash_value : int
        部分文字列のハッシュ値
    """
    length = r - l + 1
    res = hf[r] - (hf[l - 1] * pow_list[length]) % mod
    res %= mod
    return res


def solve(N: int, Q: int, S: str, queries: List[Tuple[int, int]]) -> List[str]:
    """
    与えられた文字列 S と Q 個のクエリに対して、
    各部分文字列が回文かどうかを判定する。

    Parameters
    ----------
    N : int
        文字列長
    Q : int
        クエリ数
    S : str
        対象文字列
    queries : List[Tuple[int, int]]
        各クエリ (L, R) の配列

    Returns
    -------
    results : List[str]
        各クエリに対する "Yes" または "No" の結果
    """
    MOD1, MOD2 = 1_000_000_007, 1_000_000_009
    BASE1, BASE2 = 1_000_003, 1_000_033

    pow1, hf1, hr1 = build_hash(S, BASE1, MOD1)
    pow2, hf2, hr2 = build_hash(S, BASE2, MOD2)

    results: List[str] = []
    for L, R in queries:
        fh1 = get_sub_hash(hf1, pow1, L, R, MOD1)
        fh2 = get_sub_hash(hf2, pow2, L, R, MOD2)

        revL = N - R + 1
        revR = N - L + 1
        rh1 = get_sub_hash(hr1, pow1, revL, revR, MOD1)
        rh2 = get_sub_hash(hr2, pow2, revL, revR, MOD2)

        if fh1 == rh1 and fh2 == rh2:
            results.append("Yes")
        else:
            results.append("No")

    return results


# =========================
# メイン処理
# =========================
def main() -> None:
    data = sys.stdin.buffer.read().split()
    N = int(data[0])
    Q = int(data[1])
    S = data[2].decode()
    queries: List[Tuple[int, int]] = [(int(data[i]), int(data[i + 1])) for i in range(3, 3 + 2 * Q, 2)]

    start_time = time.perf_counter()
    results = solve(N, Q, S, queries)
    end_time = time.perf_counter()

    sys.stdout.write("\n".join(results) + "\n")

    # 計測情報（stderr）
    mem_mb = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss / 1024
    sys.stderr.write(f"Time: {(end_time - start_time) * 1000:.3f} ms\n")
    sys.stderr.write(f"Memory: {mem_mb:.2f} MB\n")


if __name__ == "__main__":
    main()
# ```

# ---

# ### 実装の特徴

# 1. **二重ローリングハッシュ**

#    * `MOD1, MOD2` と `BASE1, BASE2` を変えて衝突率をほぼゼロに。
#    * Python の `int` は任意精度なので BigInt 相当。

# 2. **高速入力**

#    * `sys.stdin.buffer.read()` で一括読み取り。
#    * 文字列部分は `.decode()`、数字は `int()` に変換。

# 3. **O(1) 判定**

#    * 各クエリでの回文判定は定数時間。
#    * 全体計算量は `O(N + Q)`。

# 4. **型ヒント付き**

#    * Pylance などの型チェッカでもエラーなく通る。
#    * 関数引数・戻り値すべて明示。

# 5. **時間・メモリ計測**

#    * `time.perf_counter()` で高精度時間計測。
#    * `resource.getrusage` で最大常駐メモリ（MB）を計測。
